falcon/hooks.py

```diff
@@ -13,23 +13,41 @@
 # limitations under the License.
 
 """Hook decorators."""
+from __future__ import annotations
 
 from functools import wraps
 from inspect import getmembers
 from inspect import iscoroutinefunction
 import re
+from typing import Any
+from typing import Awaitable
+from typing import Callable
+from typing import Dict
+from typing import List
+from typing import Tuple
+from typing import TYPE_CHECKING
+from typing import Union
 
 from falcon.constants import COMBINED_METHODS
 from falcon.util.misc import get_argnames
 from falcon.util.sync import _wrap_non_coroutine_unsafe
 
+if TYPE_CHECKING:
+    import falcon as wsgi
+    from falcon import asgi
 
 _DECORABLE_METHOD_NAME = re.compile(
     r'^on_({})(_\w+)?$'.format('|'.join(method.lower() for method in COMBINED_METHODS))
 )
 
+SynchronousResource = Callable[..., Any]
+AsynchronousResource = Callable[..., Awaitable[Any]]
+Resource = Union[SynchronousResource, AsynchronousResource]
 
-def before(action, *args, is_async=False, **kwargs):
+
+def before(
+    action: Resource, *args: Any, is_async: bool = False, **kwargs: Any
```

Comment:
Referencing my comment above, `action` here would be `Union[Action, type]`. I realize that putting `type` in there makes this hint very lax, but there's nothing for it since it can be any resource class that follows the `on_*()` protocol.

Reply:
Perhaps it would be best to actually alias `type` as `Resource` so that it is at least more self-documenting.

Reply:
Done